automation/entry_scripts/balena-generate-ami.sh

```diff
@@ -0,0 +1,245 @@
+    [[ -n "${s3_image_url}" ]] && \
+      echo "* Removing img from S3..." && \
+      aws s3 rm "${s3_image_url}" && \
+      s3_image_url=""
+}
+
+
+aws_ebs_snapshot_cleanup() {
+    [[ -n "${ebs_snapshot_id}" ]] && \
+      echo "* Removing snapshot from ebs..." && \
+      aws ec2 delete-snapshot --snapshot-id "${ebs_snapshot_id}" && \
+      ebs_snapshot_id=""
+}
+
+
+create_aws_ebs_snapshot() {
+
+    local img=$1
+    local __snapshot_id=$2
+    local __s3_image_url=$3
+
+    local snapshot_id
+    local status
+    local wait_secs=2
+    local secs_waited=0
+    # https://github.com/koalaman/shellcheck/wiki/SC2155#correct-code-1
+    local s3_key
+    # Randomize to lower the chance of parallel builds colliding.
+    s3_key="tmp-$(basename "${img}")-${RANDOM}"
+
+    # Push to s3 and create the AMI
+    echo "* Pushing ${img} to s3://${S3_BUCKET}"
+    s3_url="s3://${S3_BUCKET}/preloaded-images/${s3_key}"
+    aws s3 cp "${img}" "${s3_url}"
+
+    import_task_id=$(aws ec2 import-snapshot \
+      --description "snapshot-${AMI_NAME}" \
+      --disk-container "Description=balenaOs,Format=RAW,UserBucket={S3Bucket=${S3_BUCKET},S3Key=preloaded-images/${s3_key}}" | jq -r .ImportTaskId)
+
+    echo "* Created a AWS import snapshot task with id ${import_task_id}. Waiting for completition... (Timeout: $IMPORT_SNAPSHOT_TIMEOUT_MINS mins)"
+    while true; do
+        status="$(get_value_from_ebs_snapshot_import_task "${import_task_id}" Status)"
+        [[ "$status" == "completed" ]] && break
+        [[ "$status" == "deleting" ]]  && \
+            error_msg="$(get_value_from_ebs_snapshot_import_task "${import_task_id}" StatusMessage)" && \
+            echo "ERROR: Error on import task id ${import_task_id}: '${error_msg}'" && exit 1
+
+        sleep $wait_secs
+        secs_waited=$((secs_waited + wait_secs))
+        mins_elapsed=$((secs_waited / 60))
+
+        # Show progress every 2 mins (120 secs)
+        [[ $((secs_waited % 120)) == 0 ]] && echo "-> Mins elapsed: $mins_elapsed. Progress: $(get_value_from_ebs_snapshot_import_task "${import_task_id}" Progress)%"
+        [[ "$mins_elapsed" -ge "$IMPORT_SNAPSHOT_TIMEOUT_MINS" ]] && echo "ERROR: Timeout on import snapshot taksk id ${import_task_id}" && exit 1
+    done
+
+    snapshot_id=$(aws ec2 describe-import-snapshot-tasks --import-task-ids "${import_task_id}" | jq -r '.ImportSnapshotTasks[].SnapshotTaskDetail.SnapshotId')
+    echo "* AWS import snapshot task complete. SnapshotId: ${snapshot_id}"
+
+    eval "$__snapshot_id='${snapshot_id}'"
+    eval "$__s3_image_url='${s3_url}'"
+}
+
+
+create_aws_ami() {
+
+    local snapshot_id=$1
+    local __retvalue=$2
+    local image_id
+
+    echo "Checking for AMI name conflicts"
+    existing_image_id=$(aws ec2 describe-images \
+        --filters "Name=name,Values=${AMI_NAME}" \
+        --query 'Images[*].[ImageId]' \
+        --output text)
+
+    if [ -n "${existing_image_id}" ]; then
+        echo "Cleaning up ${existing_image_id}"
```

Comment:
I would probably just exit in error instead of removing an image already deployed. In general, the cleanup functions that remove images are fine in case of errors or process interruptions, but once the image is deployed only newer versions should be pushed.

Reply:
True, remains of development mode